src/rt-app_taskgroups.c

```diff
@@ -0,0 +1,365 @@
+
+	tasks = fopen(path, "we");
+	if (!tasks) {
+		perror("fopen");
+		goto error;
+	}
+
+	if (fprintf(tasks, "%d", tid) < 0) {
+		perror("fprintf");
+		goto error;
+	}
+
+	if (fclose(tasks)) {
+		perror("fclose");
+		goto error;
+	}
+
+	ret = 0;
+error:
+	free(path);
+
+	return ret;
+}
+
+static int cgroup_mkdir(const char *name, int *offset)
+{
+	char *dir = NULL, *path = NULL, sep;
+	int i = 0, ret = 0, y;
+
+	*offset = -1;
+
+	if (!name || !strcmp(name, "/"))
+		goto error;
+
+	dir = strdup(name);
+	if (!dir) {
+		log_error("cannot duplicate string [%s]", name);
+		ret = -1;
+		goto error;
+	}
+
+	path = malloc(FILENAME_MAX);
+	if (!path) {
+		perror("malloc");
+		ret = -1;
+		goto error;
+	}
+
+	do {
+		y = i;
+
+		while (dir[i] == '/')
+			i++;
+
+		while (dir[i] != '\0' && dir[i] != '/')
+			i++;
+
+		sep = dir[i];
+		dir[i] = '\0';
+
+		snprintf(path, FILENAME_MAX, "%s%s", tg_ctrl.mount_point, dir);
+
+		ret = mkdir(path, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+		if (ret) {
+			switch (errno) {
+			case EEXIST:
+				log_debug("cgroup [%s] exists, continue ...", dir);
+				ret = 0;
+				goto next;
+			default:
+				log_error("cgroup [%s] unhandled error (%s)", dir, strerror(errno));
+				goto error;
+			}
+		} else if (*offset == -1) {
+			*offset = y;
+		}
+
+		log_debug("cgroup [%s] created", dir);
+next:
+		dir[i] = sep;
+	} while (dir[i]);
+error:
+	free(dir);
+	free(path);
+	return ret;
+}
+
+static int cgroup_rmdir(char *name, int offset)
+{
+	char *dir = NULL, *path = NULL, *sep;
+	int ret = 0;
+
+	if (offset < 0)
+		goto error;
+
+	dir = strdup(name);
+	if (!dir) {
+		log_error("cannot duplicate string [%s]", name);
+		ret = -1;
+		goto error;
+	}
+
+	path = malloc(FILENAME_MAX);
+	if (!path) {
+		perror("malloc");
+		ret = -1;
+		goto error;
+	}
+
+	while (1) {
+		snprintf(path, FILENAME_MAX, "%s%s", tg_ctrl.mount_point, dir);
+
+		ret = rmdir(path);
+		if (ret) {
+			switch (errno) {
+			case ENOTEMPTY:
+				log_debug("cgroup [%s] not empty, continue ...", dir);
+				ret = 0;
+				break;
+			case EBUSY:
+				log_debug("cgroup [%s] is busy, continue ...", dir);
+				ret = 0;
+				break;
+			case ENOENT:
+				log_debug("cgroup [%s] doesn't exist, continue ...", dir);
+				ret = 0;
+				break;
+			default:
+				log_error("cgroup [%s] unhandled error (%s)", dir, strerror(errno));
+				break;
+			}
+			break;
+		}
+
+		log_debug("cgroup [%s] removed", dir);
+
+		sep = strrchr(dir, '/');
+		if (!sep || sep <= dir + offset)
+			break;
+		*sep = '\0';
+	}
+error:
+	free(dir);
+	free(path);
+	return ret;
+}
+
+static void create_taskgroup(taskgroup_t *tg)
+{
+	if (!tg)
+		return;
+
+	log_debug("create taskgroup [%s]", tg->name);
+
+	if (cgroup_mkdir(tg->name, &tg->offset)) {
+		log_critical("cannot create taskgroup [%s]", tg->name);
+		exit(EXIT_FAILURE);
+	}
+}
+
+void create_taskgroups(rtapp_options_t *opts)
+{
+	int i, y;
+
+	for (i = 0; i < opts->nthreads; i++) {
+		thread_data_t *tdata = &opts->threads_data[i];
```

Comment:
Instead of parsing the data of all phases of all threads, why don't you create a resource per taskgroup ?
And you create the task group the 1st time you parse it

Reply:
I thought similar - we could use an ordered array of taskgroup objects (created like resources) which would simplify mount point creation/destruction and cleanup, plus potentially allocate less memory if you just point each phase at one of the allocated taskgroup objects.

Reply:
That's doable. For an ordered array I would have to pre-parse the thread (and phase) data to allocate and order the array to be able to assign ->taskgroup pointers for threads and phases in the normal parsing of the thread (and phase) data.
An ordered array has the advantage that I could get rid of the 'taskgroup->offset' ownership mechanism.